tools/walletextension/wallet_extension.go

```diff
@@ -159,6 +162,165 @@ func (w *WalletExtension) SubmitViewingKey(address gethcommon.Address, signature
 	return nil
 }
 
+// GenerateAndStoreNewUser generates new key-pair and userID, stores it in the database and returns hex encoded userID and error
+func (w *WalletExtension) GenerateAndStoreNewUser() (string, error) {
+	// generate new key-pair
+	viewingKeyPrivate, err := crypto.GenerateKey()
+	viewingPrivateKeyEcies := ecies.ImportECDSA(viewingKeyPrivate)
+	if err != nil {
+		w.Logger().Error(fmt.Sprintf("could not generate new keypair: %s", err))
+		return "", err
+	}
+
+	// create UserID and store it in the database with the private key
+	userID := calculateUserID(viewingKeyPrivate)
+	err = w.storage.AddUser(userID, crypto.FromECDSA(viewingPrivateKeyEcies.ExportECDSA()))
+	if err != nil {
+		w.Logger().Error(fmt.Sprintf("failed to save user to the database: %s", err))
+		return "", err
+	}
+
+	return hex.EncodeToString(userID), nil
+}
+
+// AddAddressToUser checks if message is in correct format and if signature is valid. If all checks pass we save address and signature against userID
+func (w *WalletExtension) AddAddressToUser(hexUserID string, message string, signature []byte) error {
+	// parse the message to get userID and account address
+	messageUserID, messageAddressHex, err := getUserIDAndAddressFromMessage(message)
+	if err != nil {
+		w.Logger().Error(fmt.Errorf("submitted message (%s) is not in the correct format", message).Error())
+		return err
+	}
+
+	// check if userID corresponds to the one in the message and check if the length of hex encoded userID is correct
+	if hexUserID != messageUserID || len(messageUserID) != common.MessageUserIDLen {
+		w.Logger().Error(fmt.Errorf("submitted message (%s) is not in the correct format", message).Error())
+		return errors.New("userID from message does not match userID from request")
+	}
+
+	// Check if the signature is valid
```

Comment:
where is the signature checked?

Reply:
I checked that a few lines below.. I check if it was signed with the same address as is mentioned in the message
`if addressFromSignature != addressFromMessage {`

Reply:
Added ecdsa.Verify (first three things were already in the code)